apps/examples/binary_update/binary_update_main.c

```diff
@@ -62,6 +94,133 @@ static void print_binary_info_list(binary_info_list_t *binary_info_list)
+		} else {
+			printf("Success to load valid higher version binary.\n");
+		}
+	} else { //DOWNLOAD_INVALID_BIN
+		if (strncmp(pre_bin_info->dev_path, cur_bin_info->dev_path, sizeof(pre_bin_info->dev_path) != 0)) {
+			fail_cnt++;
+			printf("Warning! Load invalid binary.\n");
+		} else {
+			printf("No update with invalid binary.\n");
+		}
+	}
+}
+
+static void binary_update_write_binary(char *devname, int condition, uint32_t crc_hash, binary_header_t *header_data)
+{
+	int fd;
+	int ret;
+	fd = open(devname, O_WRONLY);
+	if (fd < 0) {
+		printf("Failed to open %s: %d, errno: %d\n", devname, ret, get_errno());
+		return;
+	}
+
+	if (condition == DOWNLOAD_VALID_BIN) {
+		/* Write new crc */
+		ret = write(fd, (FAR uint8_t *)&crc_hash, CHECKSUM_SIZE);
+		if (ret != CHECKSUM_SIZE) {
+			printf("Failed to read %s: %d\n", devname, ret);
+			close(fd);
+			return;
+		}
+		printf("Download to valid binanry\n");
+	} else {
+		/*Dose not write new crc. jump to header information. */
+		ret = lseek(fd, CHECKSUM_SIZE, SEEK_SET);
+		if (ret == ERROR) {
+			printf("Failed to lseek %s\n", devname);
+			close(fd);
+			return;
+		}
+		printf("Download to invalid binanry\n");
+	}
+
+	/* Write new version. */
+	ret = write(fd, (FAR uint8_t *)header_data, sizeof(binary_header_t));
+	if (ret != sizeof(binary_header_t)) {
+		printf("Failed to write %s: %d\n", devname, ret);
+		close(fd);
+		return;
+	}
+
+	close(fd);
+}
+
+static void binary_update_download_binary(char *devname, int condition)
+{
+	int fd;
+	int ret;
+	int file_size;
+	int read_size;
+	uint32_t crc_hash = 0;
+	uint8_t crc_buffer[CRC_BUFFER_SIZE];
+	binary_header_t header_data;
+
+	printf("\n** Binary Update Download [%s] %s case test.\n", devname, condition == DOWNLOAD_VALID_BIN ? "Valid" : "Invalid");
+
+	fd = open(devname, O_RDONLY);
+	if (fd < 0) {
+		fail_cnt++;
+		printf("Failed to open %s: %d, errno: %d\n", devname, ret, get_errno());
+		return;
+	}
+
+	/* Jump to header information. */
+	ret = lseek(fd, CHECKSUM_SIZE, SEEK_SET);
+	if (ret == ERROR) {
+		printf("Failed to lseek %s\n", devname);
+		close(fd);
+		return;
+	}
+
+	/* Read the binary header. */
+	ret = read(fd, (FAR uint8_t *)&header_data, sizeof(binary_header_t));
+	if (ret != sizeof(binary_header_t)) {
+		printf("Failed to read %s: %d\n", devname, ret);
+		close(fd);
+		return;
+	}
+
+	/* Version update */
+	snprintf(header_data.bin_ver, BIN_VER_MAX, "%u", new_version);
+
+	/* Version is updated to create a new crc. */
+	if (condition == DOWNLOAD_VALID_BIN) {
+		crc_hash = crc32part((uint8_t *)&header_data, header_data.header_size, crc_hash);
+		file_size = header_data.bin_size;
+
+		while (file_size > 0) {
+			read_size = file_size < CRC_BUFFER_SIZE ? file_size : CRC_BUFFER_SIZE;
+			ret = read(fd, (FAR uint8_t *)crc_buffer, read_size);
+			if (ret != read_size) {
+				printf("Failed to read %s: %d\n", devname, ret);
+				close(fd);
+				return;
+			}
+			crc_hash = crc32part(crc_buffer, read_size, crc_hash);
```

Comment:
if crc32part fails, it will return the negative value.
Please check the return for validation.

Reply:
no. crc32part returns only positive value. so we don't know crc32part failed.